table/update_schema.go

```diff
@@ -0,0 +1,834 @@
+			us.errors = append(us.errors, err)
+		}
+	}
+
+	if len(us.errors) > 0 {
+		return fmt.Errorf("validation failed with %d errors: %v", len(us.errors), us.errors)
+	}
+	return nil
+}
+
+// Build validates and constructs the final schema
+func (us *UpdateSchema) Build() (*iceberg.Schema, error) {
+	// First validate all operations
+	if err := us.Validate(); err != nil {
+		return nil, err
+	}
+
+	// No operations means no changes
+	if len(us.operations) == 0 {
+		return us.schema, nil
+	}
+
+	// Create builder and apply operations
+	builder := &schemaBuilder{
+		deletes:      make(map[int]struct{}),
+		updates:      make(map[int]*iceberg.NestedField),
+		adds:         make(map[int][]*iceberg.NestedField),
+		moves:        make(map[int][]moveReq),
+		lastColumnID: us.lastColumnID,
+		baseSchema:   us.schema,
+		settings: &validationSettings{
+			allowIncompatibleChanges: us.allowIncompatibleChanges,
+			caseSensitive:            us.caseSensitive,
+		},
+	}
+
+	// Apply operations in order
+	for _, op := range us.operations {
+		if err := op.Apply(builder); err != nil {
+			return nil, fmt.Errorf("failed to apply operation %s: %w", op.String(), err)
+		}
+	}
+
+	// Build final schema
+	return us.buildFinalSchema(builder)
+}
+
+// Commit validates, builds, and commits the schema changes
+func (us *UpdateSchema) Commit() error {
+	updates, requirements, err := us.CommitUpdates()
+	if err != nil {
+		return err
+	}
+
+	if len(updates) == 0 {
+		return nil
+	}
+
+	return us.txn.apply(updates, requirements)
+}
+
+// CommitUpdates returns the updates and requirements needed
+func (us *UpdateSchema) CommitUpdates() ([]Update, []Requirement, error) {
+	// If there are no operations, return nil updates and requirements
+	if len(us.operations) == 0 {
+		return nil, nil, nil
+	}
+
+	newSchema, err := us.Build()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check if equivalent schema exists
+	existingSchemaID := us.findExistingSchemaInTransaction(newSchema)
+
+	var updates []Update
+	var requirements []Requirement
+
+	// for Commit Contention
+	requirements = append(requirements, AssertCurrentSchemaID(us.schema.ID))
+
+	if existingSchemaID == nil {
+		// Get the final lastColumnID from the builder
+		builder := &schemaBuilder{
+			deletes:      make(map[int]struct{}),
+			updates:      make(map[int]*iceberg.NestedField),
+			adds:         make(map[int][]*iceberg.NestedField),
+			moves:        make(map[int][]moveReq),
+			lastColumnID: us.lastColumnID,
+			baseSchema:   us.schema,
+			settings: &validationSettings{
+				allowIncompatibleChanges: us.allowIncompatibleChanges,
+				caseSensitive:            us.caseSensitive,
+			},
+		}
+
+		for _, op := range us.operations {
+			if err := op.Apply(builder); err != nil {
+				return nil, nil, fmt.Errorf("failed to recompute lastColumnID: %w", err)
+			}
+		}
+		updates = append(updates,
+			NewAddSchemaUpdate(newSchema, builder.lastColumnID, false),
+			NewSetCurrentSchemaUpdate(newSchema.ID),
+		)
```

Comment:
This seems like a waste to have to *rebuild* the schema again. Can we somehow store/update the `lastColumnID` after calling `Build()` rather than redo the build here.